src/Orleans.Core/ApplicationParts/AssemblyAttributeFeatureProvider.cs

```diff
@@ -0,0 +1,53 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Orleans.Metadata;
+
+namespace Orleans.ApplicationParts
+{
+    /// <summary>
+    /// Populates a feature using assembly-level attributes which implement <see cref="IFeaturePopulator{TFeature}"/>.
+    /// </summary>
+    /// <typeparam name="TFeature">The feature type.</typeparam>
+    public sealed class AssemblyAttributeFeatureProvider<TFeature> : IApplicationFeatureProvider<TFeature>
+    {
+        /// <summary>
+        /// Returns <see langword="true"/> if this instance is equivalent to the provided instance, <see langword="false"/> otherwise.
+        /// </summary>
+        /// <param name="other">The other instance.</param>
+        /// <returns>
+        /// <see langword="true"/> if this instance is equivalent to the provided instance, <see langword="false"/> otherwise.
+        /// </returns>
+        public bool Equals(AssemblyAttributeFeatureProvider<TFeature> other) => true;
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            return obj.GetType() == this.GetType();
+        }
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+        {
+            return typeof(AssemblyAttributeFeatureProvider<TFeature>).GetHashCode();
+        }
+
+        /// <inheritdoc />
+        public void PopulateFeature(IEnumerable<IApplicationPart> parts, TFeature feature)
+        {
+            foreach (var part in parts.OfType<AssemblyPart>())
+            {
+                var attributes = part.Assembly.GetCustomAttributes(true);
```

Comment:
since we are using a well known attribute now, you can get this attribute directly instead of getting all and then trying to cast to `FeaturePopulatorAttribute`. I believe that approach is also safer against assembly load exception if some attribute is in another assembly that can't be loaded